### ISSUE TITLE
Create Monorail Client Interface to Improve Testability

### ISSUE THREAD

**Issue by spiegela** [emccode/gorackhd#24]
Right now, the current design of the RackHD Monorail Client (nested structs) is not mockable in dependent applications.

In order to test the client with mocks, it would be helpful to implement each of the APIs as methods (`Nodes()`, `Catalogs()`, etc.) and create a Monorail client interface. Then we could create an interface for each of the cilent endpoints -- a Nodes interface, Catalogs interface, etc.

**Comment by spiegela** [comment]
BTW, I've started a mockable interface in a dependent app (`infrakit.rackhd`), so I'll have a bunch of this code ready, and can probably submit a PR for this once I get past deadlines.

**Comment by kacole2** [comment]
I'm going to give you write access to this repo since you are going to be relying on it heavily while Travis and myself haven't used it in a while. 

Kendrick Coleman
kendrickcoleman.com
@KendrickColeman
github.com/kacole2

**Comment by spiegela** [comment]
Sweet. Thanks. Once I have an interface and mocks fleshed out, I'll start moving them over to this repo.

**Comment by kacole2** [comment]
You've been added as an admin. Sorry for the late response. 

Kendrick Coleman
kendrickcoleman.com
@KendrickColeman
github.com/kacole2

**Comment by codenrhoden** [comment]
One thing to keep in mind is that 99.9% of the code in this repo is auto-generated. It's not something I looked into, but with the modifications you are proposing to the client, will you still be able to generate the code and have the features you need?

Are you wanting to put a wrapper in front of the auto-gen stuff? I think that's where you are heading with creating interfaces for each of the client endpoints, which sounds like the right approach to me.

**Comment by spiegela** [comment]
Yes, I saw that the client code is autogenerated using swagger API, and yes, the intent would be to wrap the generated code.

**Comment by akutz** [comment]
Good luck with that. I have experience with Swagger. And it's not good. My recollection is that at the time the generated code was very poor and you needed to actively modify it to your needs. Wrapping was insufficient since they would hide certain HTTP workflows from you.
